hack/verify-generated-protobuf.sh

```diff
@@ -1,65 +0,0 @@
-#!/usr/bin/env bash
-
-# Copyright 2015 The Kubernetes Authors.
-#
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
-# You may obtain a copy of the License at
-#
-#     http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-
-# This script checks whether updating of the packages which explicitly
-# requesting generation is needed or not. We should run
-# `hack/update-generated-protobuf.sh` if those packages are out of date.
-# Usage: `hack/verify-generated-protobuf.sh`.
-
-set -o errexit
-set -o nounset
-set -o pipefail
-
-KUBE_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
-source "${KUBE_ROOT}/hack/lib/init.sh"
-
-kube::golang::setup_env
-
-APIROOTS=$({
-  # gather the packages explicitly requesting generation
-  git grep --files-with-matches -e '// +k8s:protobuf-gen=package' cmd pkg staging | xargs -n 1 dirname
-  # add the root apimachinery pkg containing implicitly generated files (--apimachinery-packages)
-  echo staging/src/k8s.io/apimachinery/pkg
-} | sort | uniq)
-
-_tmp="${KUBE_ROOT}/_tmp"
-
-cleanup() {
-  rm -rf "${_tmp}"
-}
-
-trap "cleanup" EXIT SIGINT
-
-cleanup
-for APIROOT in ${APIROOTS}; do
-  mkdir -p "${_tmp}/${APIROOT}"
-  cp -a "${KUBE_ROOT}/${APIROOT}"/* "${_tmp}/${APIROOT}/"
-done
-
-KUBE_VERBOSE=3 "${KUBE_ROOT}/hack/update-generated-protobuf.sh"
-for APIROOT in ${APIROOTS}; do
-  TMP_APIROOT="${_tmp}/${APIROOT}"
-  echo "diffing ${APIROOT} against freshly generated protobuf"
-  ret=0
-  diff -Naupr -I 'Auto generated by' -x 'zz_generated.*' -x '.github' -x '.import-restrictions' "${KUBE_ROOT}/${APIROOT}" "${TMP_APIROOT}" || ret=$?
-  cp -a "${TMP_APIROOT}"/* "${KUBE_ROOT}/${APIROOT}/"
-  if [[ $ret -eq 0 ]]; then
-    echo "${APIROOT} up to date."
-  else
-    echo "${APIROOT} is out of date. Please run hack/update-generated-protobuf.sh"
```

Comment:
how is the protobug generation verified now? it seems it passes th flag `verify-only` but I can't see it used in the new function

Reply:
We make a worktree, remove all generated files, re-generate, and diff.  It's far cleaner and more consistent.  Many tools do not have a `--verify` flag.

Reply:
```
$ ./hack/verify-codegen.sh protobuf
+++ [0122 09:24:54] Generating protobufs for 66 targets
+++ [0122 09:24:54] Verifying Prerequisites....
+++ [0122 09:24:55] Building Docker image kube-build:build-aeca35f1ba-5-v1.26.0-go1.19.5-bullseye.0
+++ [0122 09:24:57] Creating data container kube-build-data-aeca35f1ba-5-v1.26.0-go1.19.5-bullseye.0
+++ [0122 09:25:10] Syncing sources to container
+++ [0122 09:25:14] Output from this container will be rsynced out upon completion. Set KUBE_RUN_COPY_OUTPUT=n to disable.
+++ [0122 09:25:14] Running build command...
+++ [0122 09:25:41] Syncing out of container
Not currently on any branch.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   staging/src/k8s.io/api/networking/v1/generated.pb.go
	modified:   staging/src/k8s.io/api/networking/v1/generated.proto

no changes added to commit (use "git add" and/or "git commit -a")
diff --git a/staging/src/k8s.io/api/networking/v1/generated.pb.go b/staging/src/k8s.io/api/networking/v1/generated.pb.go
index e9566d57e27..9c86b6c966b 100644
--- a/staging/src/k8s.io/api/networking/v1/generated.pb.go
+++ b/staging/src/k8s.io/api/networking/v1/generated.pb.go
@@ -868,114 +868,115 @@ func init() {
 }
 
 var fileDescriptor_1c72867a70a7cc90 = []byte{
-       // 1715 bytes of a gzipped FileDescriptorProto

<elided for this comment>

Generated files need to be updated
Please run 'hack/update-codegen.sh'
```

Reply:
I'm trying to find where that magic happens 😄

Reply:
verify-codegen.sh

```
_tmpdir="$(kube::realpath "$(mktemp -d -t "$(basename "$0").XXXXXX")")"
git worktree add -f -q "${_tmpdir}" HEAD
kube::util::trap_add "git worktree remove -f ${_tmpdir}" EXIT
cd "${_tmpdir}"

# Update generated code
export UPDATE_API_KNOWN_VIOLATIONS=true
hack/update-codegen.sh "$@"

# Test for diffs
diffs=$(git status --porcelain | wc -l)
```

Reply:
duh , hahah, I had an old version checked out